test/explicit-resource-management/main.test.ts

```diff
@@ -0,0 +1,120 @@
+
+import { expect } from 'chai';
+import { describe, it } from 'mocha';
+import { AbstractCursor, ChangeStream, ClientSession, GridFSBucket, MongoClient } from 'mongodb/lib/beta';
+import * as sinon from 'sinon';
+import { Readable } from 'stream';
+import { pipeline } from 'stream/promises';
+import { setTimeout } from 'timers/promises';
+
+async function setUpCollection(client: MongoClient) {
+	const collection = client.db('foo').collection<{ name: string }>('bar');
+	const documents: Array<{ name: string }> = Array.from({ length: 5 }).map(i => ({
+		name: String(i)
+	}));
+	await collection.insertMany(documents)
+	return collection;
+}
+
+describe('explicit resource management smoke tests', function () {
+	const clientSpy = sinon.spy(MongoClient.prototype, Symbol.asyncDispose);
+	const cursorSpy = sinon.spy(AbstractCursor.prototype, Symbol.asyncDispose);
+	const endSessionSpy = sinon.spy(ClientSession.prototype, Symbol.asyncDispose);
+	const changeStreamSpy = sinon.spy(ChangeStream.prototype, Symbol.asyncDispose);
+	const readableSpy = sinon.spy(Readable.prototype, Symbol.asyncDispose);
+
+	afterEach(function () {
+		clientSpy.resetHistory();
+		cursorSpy.resetHistory();
+		endSessionSpy.resetHistory();
+		changeStreamSpy.resetHistory();
+		readableSpy.resetHistory();
+	});
+
+	describe('MongoClient', function () {
+		it('can be used with await-using syntax', async function () {
```

Comment:
"can be used with await-using syntax" is somewhat vague intent-wise, e.g., to me, it means "doesn't crash when used with await-using syntax". I think we should be testing the part of the user expectation that really matters here, i.e., that `it('is cleaned up automatically when used with await-using syntax')` and assert that the corresponding `close` method is called as expected, instead of the `asyncDispose` (or in addition to it, I don't have strong opinion on whether or not to keep that part of the assertion)

what I am looking for is a set of tests that covers the expected user-facing behavior: it can be here or it can be in a different suite if this one is meant to serve a more limited purpose

Reply:
I do think there is still value in testing the feature in the context in which it will be used by our users. If, hypothetically speaking, the runtime/bundler has an issue where the way we have implemented asyncDispose is no longer sufficient to ensure proper cleanup, it would be better to find out about it in our test suite as opposed to from user reports. There is also the additional benefit of ensuring that our cleanup methods themselves adhere to the idempotency expected from the asyncDispose implementation: that if the user's code inside the context block results in our driver automatically cleaning up the resource, then calling it again implicitly at the end of a context block doesn't cause any issues.

Reply:
I disagree - it is not our responsibility to test the implementation of bundlers or runtimes in this way.  The resource management proposal is more-or-less stable and the changes being made are all minor, internal implementation details.  I mentioned this in the [spike](https://docs.google.com/document/d/1eT-7ntnX--M49OPmLO41RNe1PJ8Vore6TE2Kz3IY_cI/edit#heading=h.hxndodn23tzy).  This feature has had a year of adoption in the community and any breakage in tooling would disrupt the ecosystem enough that it would be caught quickly.

The implementation of asyncDispose is tested separately.  Our dispose methods proxy to the resources' underlying disposal method, and any behavioral requirements for `close()` or `endSession()` should be tested elsewhere.

I believe the current testing sufficient, so unless you consider additional testing a blocker to merging this PR, can we consider this resolved?  If not, can you elaborate on exactly what testing you'd like to see added before you would consider this PR unblocked?

Reply:
I believe that it is in the best interest of our users that we test any workflow that we advertise to them, i.e., we should contain tests that, at a minimum, cover the syntax we use in our documentation, and ideally mimic any end-to-end workflow that we would consider supported.

It's the same reason we write integration tests for other driver features and don't consider it sufficient to just unit test each individual component: things can break down in unexpected ways when put together no matter how thoroughly you try to define and test the individual functionalities and the contracts between them, so the best way to guarantee a smooth end-to-end experience is to test the end-to-end experience.

Based on my understanding of the purpose of this work, supported user flows concern automatic cleanup of the client, sessions, and cursors. I would like to see the happy path and main error scenarios covered for these flows. Happy path would cover successfully instantiating the resource and then, if it is stateful, checking that the cleanup happens as expected for each state (e.g., for cursors we'd want to cover instantiation, mid-iteration, and exhaustion separately). Error means we get the expected error with `using` even if the resource fails to instantiate (e.g., uri validation in the client) or fails later with an error. I think you could collaborate with the primary reviewer (@nbbeeken) on the exact lineup of test cases that can help ensure that users don't run into surprising behavior when they go to try out this feature.

Reply:
Met with @nbbeeken and I've added tests for the scenarios we discussed.

Reply:
Resolving this thread in favor of the other test-specific thread since some of the concerns have now been addressed in the docs changes and the tests